src/classes/RD2_ScheduleService.cls

```diff
@@ -909,9 +865,242 @@ public without sharing class RD2_ScheduleService {
+        public Date endDate { get; set; }
+        public Decimal installmentAmount { get; set; }
+        public Integer installmentFrequency { get; set; }
+        public String installmentPeriod { get; set; }
+        public String paymentMethod { get; set; }
+        public String recurringDonationId { get; set; }
+        public Id scheduleId { get; set; }
+        public Date startDate { get; set; }
+
+        public ActiveSchedule(RecurringDonationSchedule__c rdSchedule, npe03__Recurring_Donation__c rd) {
+            isCurrent = rdSchedule.StartDate__c <= RD2_ScheduleService.currentDate ? true : false;
+            campaignId = rdSchedule.Campaign__c;
+            campaignName = rdSchedule.Campaign__r.Name;
+            dayOfMonth = rdSchedule.InstallmentPeriod__c == RD2_Constants.INSTALLMENT_PERIOD_MONTHLY ? rdSchedule.DayOfMonth__c : null;
+            endDate = (rdSchedule.EndDate__c != null) ? rdSchedule.EndDate__c : rd.EndDate__c;
+            installmentAmount = rdSchedule.InstallmentAmount__c;
+            installmentFrequency = rdSchedule.InstallmentFrequency__c != null ? Integer.valueOf(rdSchedule.InstallmentFrequency__c) : null;
+            installmentPeriod = rdSchedule.InstallmentPeriod__c;
+            paymentMethod = rdSchedule.PaymentMethod__c;
+            recurringDonationId = rdSchedule.RecurringDonation__c;
+            scheduleId = rdSchedule.Id;
+            startDate = rdSchedule.StartDate__c;
+        }
+    }
+
+    /***
+    * @description Handles Pause Recurring Donation functionality
+    */
+    public class PauseScheduleHandler {
+
+        /***
+        * @description Simple check on the schedule to check if the "Is Pause" field is checked
+        * @param schedule Recurring Donation Schedule
+        * @return Boolean
+        */
+        public Boolean isPause(RecurringDonationSchedule__c schedule) {
+            return schedule.IsPause__c;
+        }
+
+        /***
+        * @description Indicates if the schedule is an active pause schedule
+        * @param schedule Recurring Donation Schedule
+        * @param referenceDate As of today to compare the Schedule Start and End Dates against
+        * @return Boolean
+        */
+        public Boolean isActivePause(RecurringDonationSchedule__c schedule, Date referenceDate) {
+            if (!schedule.IsPause__c) {
+                return false;
+            }
+
+            return isWithinDateRange(schedule, referenceDate);
+        }
+
+        /***
+        * @description Indicates if the schedule is active starting from the reference date
+        * @param schedule Recurring Donation schedule
+        * @param referenceDate Reference Date
+        * @return Boolean
+        */
+        private Boolean isWithinDateRange(RecurringDonationSchedule__c schedule, Date referenceDate) {
+            return schedule.EndDate__c == null
+                || (referenceDate <= schedule.EndDate__c && schedule.StartDate__c <= schedule.EndDate__c);
```

Comment:
@RadSFDeveloper The `if` condition here has the following condition, which I would assume is true 100% of the time.
`schedule.StartDate__c <= schedule.EndDate__c`

The way I read this method is:
- If EndDate is null, return true
- If EndDate is not null and the referenceDate is before or equal to the EndDate, return true (because StartDate must always be before or equal to EndDate).
- If EndDate is not null and the referenceDate is after the EndDate, return false.

Could this impact how the method is used?

Reply:
If the End Date is in past, the schedule is not returned regardless of the Start Date.
If the End Date is today or in future *and* the Start Date <= End Date, the Pause Schedule is a valid schedule and it should be returned/handled. So the above conclusion is correct, let me know if it does not work as expected.

When testing and code reviewing, change Enhanced RD page layout to add RD Schedule related list and verify schedules upon creation, and deactivation. Also, you can change the Start Date or even End Date manually to simulate current or past valid schedules. Past valid  and deactivated schedules should not be used in any schedule and installment processing. Current and future valid schedules should be handled. Current and future invalid (inactive) schedules should not be handled. Also, there is an assumption that at most one active Pause Schedule can exist.

FYI @rponti-sforg

Reply:
Also, when the RD is closed, any active schedule including the Pause Schedule is deactivated.
FYI @force2b , @npsp-reedestockton , @rponti-sforg